pallets/name-service/src/lib.rs

```diff
@@ -0,0 +1,405 @@
+            T::Hashing::hash_of(&registrant),
+        );
+
+        // Check if already Pending Register
+        ensure!(
+            !PendingRegister::<T>::contains_key(hash_user),
+            Error::<T>::AlreadyPendingRegister
+        );
+
+        // Check if already registered
+        ensure!(
+            !UsernameRecords::<T>::contains_key(username),
+            Error::<T>::NameAlreadyRegistered
+        );
+
+        PendingRegister::<T>::insert(
+            hash_user,
+            (
+                hash_address,
+                frame_system::Pallet::<T>::block_number()
+                    .saturating_add(T::RegisterWaitingPeriod::get()),
+            ),
+        );
+
+        Self::deposit_event(Event::NameQueuedForRegister);
+        Ok(())
+    }
+
+    /// Finish Register after block time has passed
+    fn do_accept_register(username: &UserName, registrant: ZkAddressType) -> DispatchResult {
+        // Username checks
+        username_validation(username).ok_or(Error::<T>::InvalidUsernameFormat)?;
+
+        let (hash_user, hash_address) = (
+            T::Hashing::hash_of(username),
+            T::Hashing::hash_of(&registrant),
+        );
+
+        // check if block number has been passed
+        let pending_register_data =
+            PendingRegister::<T>::get(hash_user).ok_or(Error::<T>::UsernameNotFound)?;
+        ensure!(
+            frame_system::Pallet::<T>::block_number() > pending_register_data.1,
+            Error::<T>::RegisterTimeNotReached
+        );
+
+        ensure!(
+            pending_register_data.0 == hash_address,
+            Error::<T>::NotOwned
+        );
+
+        // Move from pending into records
+        PendingRegister::<T>::remove(hash_user);
+        UsernameRecords::<T>::insert(username, registrant);
+
+        Self::deposit_event(Event::NameRegistered);
+        Ok(())
+    }
+
+    /// Set primary name if register and owned
+    fn try_set_primary_name(username: UserName, registrant: ZkAddressType) -> DispatchResult {
+        //check if name is registered
+        ensure!(
+            UsernameRecords::<T>::contains_key(&username),
+            Error::<T>::NotRegistered
+        );
+        // check if its owned (can unwrap because of previous checks for the key)
+        ensure!(
+            UsernameRecords::<T>::get(&username).unwrap() == registrant,
+            Error::<T>::NotOwned
+        );
+
+        // check if we already have a primary
+        if PrimaryRecords::<T>::contains_key(registrant) {
+            PrimaryRecords::<T>::mutate(registrant, |old_username| *old_username = Some(username));
+        } else {
+            PrimaryRecords::<T>::insert(registrant, username);
+        }
+
+        Self::deposit_event(Event::NameSetAsPrimary);
+        Ok(())
+    }
+
+    fn try_cancel_pending_register(
+        username: UserName,
+        registrant: ZkAddressType,
+    ) -> DispatchResult {
+        let (hash_user, hash_address) = (
```

Comment:
validate here as well.
Or you can simply make the UserName a BoundedVec with size 63